### PR TITLE
feat: loop unrolling support

### LINKED ISSUE #1965: feat: tgpu.unroll
As per #1956.

### PR DESCRIPTION
**Pull request by cieplypolar**
This PR introduces loop unrolling.

More about unroll #1956.

Closes #1965.
#2188 and #2177 should be merged before.

TODO:
- [x] forOf and unroll docs

### CODE REVIEW

File: packages/typegpu/src/tgsl/accessIndex.ts [hunk 1]
```diff
@@ -129,5 +131,12 @@ export function accessIndex(
     );
   }
 
+  if (
+    isWgslStruct(target.dataType) && isKnownAtComptime(index) &&
+    typeof index.value === 'string'
+  ) {
+    return accessProp(target, index.value);
+  }
```
Comment (iwoplaza) [inline packages/typegpu/src/tgsl/accessIndex.ts:139]: 🔥

File: packages/typegpu/tests/unroll.test.ts [hunk 2]
```diff
@@ -0,0 +1,648 @@
...
+        }
+        {
+          result += 3i;
+        }
+        return result;
+      }"
+    `);
+  });
+
+  it('warns when iterable element count is unknown at compile-time and fallbacks to regular loop', ({ root }) => {
```
Comment (iwoplaza) [inline packages/typegpu/tests/unroll.test.ts]:
```suggestion
  it('throws when iterable element count is unknown at compile-time', ({ root }) => {
```

File: packages/typegpu/tests/unroll.test.ts [hunk 3]
```diff
@@ -0,0 +1,648 @@
...
+      [Error: Resolution of the following tree failed:
+      - <root>
+      - fn*:f
+      - fn*:f()
+      - ArrayExpression: 'v1' reference cannot be used in an array constructor.
+      -----
+      Try 'vec2f(v1)' or 'arrayOf(vec2f, count)([...])' to copy the value instead.
+      -----]
+    `);
+  });
```
Comment (iwoplaza) [inline packages/typegpu/tests/unroll.test.ts]: I thought we made the ArrayExpression to allow this use of unroll?

Comment (cieplypolar) [reply packages/typegpu/tests/unroll.test.ts]: It did. I changed it few hours ago. I forgot about our agreement and I made it consistent with regular array expressions.

Comment (cieplypolar) [reply packages/typegpu/tests/unroll.test.ts]: Pointers are back

File: packages/typegpu/src/tgsl/wgslGenerator.ts [hunk 4]
```diff
@@ -1166,127 +1185,160 @@ ${this.ctx.pre}else ${alternate}`;
...
-          'Only `for (const ... of ... )` loops are supported',
-        );
-      }
+            // TODO: replace it with tinyest traversal
+            if (
+              blocks[0]?.includes('break') || blocks[0]?.includes('continue')
+            ) {
+              throw new WgslTypeError(
+                'Cannot unroll loop containing `break` or `continue`',
+              );
```
[the remark this answers is not in the document]
Comment (cieplypolar) [reply packages/typegpu/src/tgsl/wgslGenerator.ts]: Nice edge case copilot

File: packages/typegpu/tests/unroll.test.ts [hunk 5]
```diff
@@ -0,0 +1,855 @@
...
+          // unrolled iteration #1, 'boo' is 'Boid()'
+          {
+            let baz = (&b2);
+            res = ((res + (*baz).vel) + Boid().vel);
+          }
+        }
+        return res;
+      }"
+    `);
+  });
```
Comment (aleksanderkatan) [inline packages/typegpu/tests/unroll.test.ts:181]: this behavior is not consistent with js

```ts
  it('unrolls array expression of complex types', () => {
    const Boid = d.struct({
      pos: d.vec2i,
      vel: d.vec2f,
    });

    const f = () => {
      'use gpu';
      let b = Boid();
      for (const foo of tgpu.unroll([Boid()])) {
        foo.pos = d.vec2i();
        b = foo;
      }

      return b;
    };

    expect(tgpu.resolve([f])).toMatchInlineSnapshot(`
      "struct Boid {
        pos: vec2i,
        vel: vec2f,
      }

      fn f() -> Boid {
        var b = Boid();
        // unrolled iteration #0, 'foo' is 'Boid()'
        {
          Boid().pos = vec2i();
          b = Boid();
        }
        return b;
      }"
    `);
  });
```

File: packages/typegpu/src/tgsl/wgslGenerator.ts [hunk 6]
```diff
@@ -47,12 +47,13 @@ import type { ShaderGenerator } from './shaderGenerator.ts';
 import { createPtrFromOrigin, implicitFrom, ptrFn } from '../data/ptr.ts';
 import { RefOperator } from '../data/ref.ts';
 import { constant } from '../core/constant/tgpuConstant.ts';
+import { UnrolledIterable as UnrollableIterable } from '../../src/core/unroll/tgpuUnroll.ts';
```
Comment (iwoplaza) [inline packages/typegpu/src/tgsl/wgslGenerator.ts]:
```suggestion
import { UnrolledIterable as UnrollableIterable } from '../core/unroll/tgpuUnroll.ts';
```

File: packages/typegpu/src/tgsl/wgslGenerator.ts [hunk 7]
```diff
@@ -1170,114 +1175,128 @@ ${this.ctx.pre}else ${alternate}`;
...
-          );
-        }
+        const originalLoopVarName = loopVar[1];
+        const blockified = blockifySingleStatement(body);
 
-        const iterableDataType = iterableSnippet.dataType;
-        let elementCountSnippet: Snippet;
-        let elementType = elementSnippet.dataType;
+        if (shouldUnroll) {
+          if ((stitch`${elementCountSnippet}`).includes('arrayLength')) {
```
Comment (iwoplaza) [inline packages/typegpu/src/tgsl/wgslGenerator.ts]: I think this could be checked with `isComptimeKnown`

File: packages/typegpu/src/types.ts [hunk 8]
```diff
@@ -161,6 +161,8 @@ export interface ItemStateStack {
   readSlot<T>(slot: TgpuSlot<T>): T | undefined;
   getSnippetById(id: string): Snippet | undefined;
   defineBlockVariable(id: string, snippet: Snippet): void;
+  setBlockExternals(externals: Record<string, Snippet>): void;
+  clearBlockExternals(): void;
```
Comment (aleksanderkatan) [inline packages/typegpu/src/types.ts]: duplicated lines

File: packages/typegpu/tests/unroll.test.ts [hunk 9]
```diff
@@ -0,0 +1,824 @@
...
+          const boo = 1;
+        }
+        // unrolled iteration #1, 'foo2' is '2'
+        {
+          const boo = 2;
+        }
+      }"
+    `);
+  });
+
```
Comment (aleksanderkatan) [inline packages/typegpu/tests/unroll.test.ts:93]: free test

```ts
  it('unrolls correctly an empty loop', () => {
    const arr: number[] = [];
    const f = () => {
      'use gpu';
      let a = 0;
      for (const foo of tgpu.unroll(arr)) {
        a += 1;
      }
      return a;
    };

    expect(tgpu.resolve([f])).toMatchInlineSnapshot(`
      "fn f() -> i32 {
        var a = 0;
        return a;
      }"
    `);
  });
```

File: packages/typegpu/tests/unroll.test.ts [hunk 10]
```diff
@@ -0,0 +1,824 @@
...
+            {
+              continue;
+            }
+          }
+        }
+      }"
+    `);
+  });
+
+  it('unrolling flag is set correclty', () => {
```
Comment (iwoplaza) [inline packages/typegpu/tests/unroll.test.ts]:
```suggestion
  it('unrolling flag is set correctly', () => {
```

File: packages/typegpu/tests/tgsl/wgslGenerator.test.ts [hunk 11]
```diff
@@ -1112,8 +1145,8 @@ describe('wgslGenerator', () => {
       const res = wgslGenerator.expression(
         // deno-fmt-ignore: it's better that way
         (
-          astInfo.ast?.body[1][0] as tinyest.Const
-        )[2] as unknown as tinyest.Expression,
+            astInfo.ast?.body[1][0] as tinyest.Const
+          )[2] as unknown as tinyest.Expression,
```
Comment (iwoplaza) [inline packages/typegpu/tests/tgsl/wgslGenerator.test.ts]:
```suggestion
          astInfo.ast?.body[1][0] as tinyest.Const
        )[2] as unknown as tinyest.Expression,
```

File: packages/typegpu/src/tgsl/wgslGenerator.ts [hunk 12]
```diff
@@ -1170,114 +1175,128 @@ ${this.ctx.pre}else ${alternate}`;
...
+            !wgsl.isNaturallyEphemeral(elements[0]?.dataType)
+          ) {
+            throw new WgslTypeError(
+              'Cannot unroll loop. The elements of iterable are emphemeral but not naturally ephemeral.',
+            );
+          }
+
+          const blocks = elements
+            .map((e, i) =>
+              `${this.ctx.pre}// unrolled iteration #${i}, '${originalLoopVarName}' is '${stitch`${e}`}'\n${this.ctx.pre}${
```
Comment (aleksanderkatan) [inline packages/typegpu/src/tgsl/wgslGenerator.ts]: Are we sure that `e` will fit in one line? maybe we should .replaceAll("\n", "")?

Comment (cieplypolar) [reply packages/typegpu/src/tgsl/wgslGenerator.ts]: good point
